---
title: "BIO-463: Final Project - Group 5"
author: "Marija Zelic, SCIPER: 371272"
date: "May 2023"
output:
  html_document:
    theme: paper
#paper readable
    highlight: monochrome
    code_folding: hide
    toc: true
    toc_depth: 2
#    toc_float: true
    number_sections: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(warning = FALSE, message = FALSE) 
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(cache =TRUE)
knitr::opts_chunk$set(fig.width=16, fig.height=10, figure.align = "center") 
rm(list = ls())
```

```{r libraries, include=FALSE}
library(tximport)
library(GenomicFeatures)
library(DESeq2)
library(org.Hs.eg.db)
library(RColorBrewer)
library(ggrepel)
library(ggplot2)
library(pheatmap)
```

Important to note that they use this tximport function.

```{r loadata, include=FALSE}

# Let's build objects needed to import the transcripts
setwd("/Users/marijazelic/Downloads/FinalProject_GB/GSE180371_RAW")
txdb = makeTxDbFromGFF("/Users/marijazelic/Downloads/FinalProject_GB/gencode.v46lift37.basic.annotation.gff3.gz")

k = keys(txdb, keytype="TXNAME")
tx2gene = ensembldb::select(txdb, k, "GENEID", "TXNAME")

# The salmon files don't have decimal in the ensembl IDs, so I manually cut them off
a = gsub("\\..*", "", tx2gene[,1])
b = gsub("\\..*", "", tx2gene[,2])
c = cbind(a, b)
colnames(c) = colnames(tx2gene)
tx2gene = as.data.frame(c)

dir = "/Users/marijazelic/Downloads/FinalProject_GB/GSE180371_RAW"
files = list.files(dir)
names(files) = c("1309", "1310", "1311", "1312", "1313", "1314", "1315", "1316", "1317", "1318", "1319", "1320", "1321", "1322", "1323", "1324", "1325", "1326")
# Let's import the quantifications
mat_gse = tximport(files = files,
           type = "salmon", tx2gene=tx2gene, geneIdCol = "Name",
           txIdCol = NULL,
           abundanceCol = "TPM",
           countsCol = "NumReads",
           lengthCol = "EffectiveLength")
```

```{r deseq21, include=TRUE}

# Make a sample table that will serve for design formula
sampleTable = as.data.frame(rep(c("Control", "Hypoxia", "Vadadustat"), 6))
colnames(sampleTable) = as.factor("condition")
rownames(sampleTable) = colnames(mat_gse$counts)

# Run DESeq2
dds = DESeqDataSetFromTximport(mat_gse, sampleTable, ~condition)
dds = DESeq(dds)
results = results(dds, contrast = c("condition", "Vadadustat", "Hypoxia"))

# Serves for mapping ENSEML gene names to gene SYMBOLS
results$symbol = mapIds(org.Hs.eg.db, keys=rownames(results), keytype = "ENSEMBL", column = "SYMBOL")

# For upregulated/downregulated
id = which(results$padj < 0.05)
results1 = results[id, ]

# For plotting
results$diffexpressed[results$padj < 0.05] = "significant"
results$diffexpressed[results$padj > 0.05] = "not significanat"
indexes = which(is.na(results$diffexpressed))

# Remove all NA rows from results
results = results[-indexes, ]

# For heatmap
# Select columns for Vadadustat vs. Hypoxia
columns = c("1310", "1313", "1316", "1319", "1322", "1325", "1311", "1314", "1317", "1320", "1323", "1326")

v_h_data = as.data.frame(mat_gse$counts[id, columns])
v_h_data = v_h_data[rowSums(v_h_data != 0) > 0, ]
```

```{r pheatmapI, include=TRUE}
custom_palette = colorRampPalette(c("red", "black", "green"))(3)
pheatmap(v_h_data, cluster_rows = F, scale="row", color=custom_palette)
```

```{r volcanoplotI, include=TRUE}

mycolors = c("black", "red")
vp1 = ggplot(data=results, aes(x=as.numeric(log2FoldChange), y=-log10(as.numeric(padj)), col=diffexpressed, label=symbol)) +
  geom_point() +
  geom_text_repel(data = subset(results, padj <= 0.05), aes(x=as.numeric(log2FoldChange), y=-log10(as.numeric(padj)), label=symbol)) + labs(title="Vadadustat vs. Hypoxia", y="-log10(adjP-value)", x="Log2FoldChange") +
  scale_colour_manual(values = mycolors)
vp1
ggsave('~/saved_image.png', plot = last_plot())

```

```{r deseq22, include=TRUE}

# Run DESeq2
dds = DESeqDataSetFromTximport(mat_gse, sampleTable, ~condition)
dds = DESeq(dds)
results = results(dds, contrast = c("condition", "Hypoxia", "Control"))

# Serves for mapping ENSEML gene names to gene SYMBOLS
results$symbol = mapIds(org.Hs.eg.db, keys=rownames(results), keytype = "ENSEMBL", column = "SYMBOL")

# For upregulated/downregulated
id = which(results$padj < 0.05)
results2 = results[id, ]

# For plotting
results$diffexpressed[results$padj <= 0.05] = "significant"
results$diffexpressed[results$padj > 0.05] = "not significanat"
indexes = which(is.na(results$diffexpressed))

# Remove all NA rows from results
results = results[-indexes, ]
```

```{r volcanoplotII, include=TRUE}

mycolors = c("black", "red")
vp1 = ggplot(data=results, aes(x=as.numeric(log2FoldChange), y=-log10(as.numeric(padj)), col=diffexpressed, label=symbol)) +
  geom_point() +
  geom_text_repel(data = subset(results, padj <= 0.05), aes(x=as.numeric(log2FoldChange), y=-log10(as.numeric(padj)), label=symbol)) + labs(title="Hypoxia vs. Control", y="-log10(adjP-value)", x="Log2FoldChange") +
  scale_colour_manual(values = mycolors)
vp1
ggsave('~/saved_image1.png', plot = last_plot())

```

```{r deseq23, include=TRUE}

# Run DESeq2
dds = DESeqDataSetFromTximport(mat_gse, sampleTable, ~condition)
dds = DESeq(dds)
results = results(dds, contrast = c("condition", "Vadadustat", "Control"))

# Serves for mapping ENSEML gene names to gene SYMBOLS
results$symbol = mapIds(org.Hs.eg.db, keys=rownames(results), keytype = "ENSEMBL", column = "SYMBOL")

# For upregulated/downregulated
id = which(results$padj < 0.05)
results3 = results[id, ]

# For plotting
results$diffexpressed[results$padj <= 0.05] = "significant"
results$diffexpressed[results$padj > 0.05] = "not significanat"
indexes = which(is.na(results$diffexpressed))

# Remove all NA rows from results
results = results[-indexes, ]
```

```{r volcanoplotIII, include=TRUE}

mycolors = c("black", "red")
vp1 = ggplot(data=results, aes(x=as.numeric(log2FoldChange), y=-log10(as.numeric(padj)), col=diffexpressed, label=symbol)) +
  geom_point() +
  geom_text_repel(data = subset(results, padj <= 0.05), aes(x=as.numeric(log2FoldChange), y=-log10(as.numeric(padj)), label=symbol)) + labs(title="Vadadustat vs. Control", y="-log10(adjP-value)", x="Log2FoldChange") +
  scale_colour_manual(values = mycolors)
vp1
ggsave('~/saved_image2.png', plot = last_plot())

```

```{r updown, include=TRUE}
results1$diffexpressed[results1$log2FoldChange < 0.0] = "Downregulated"
results1$diffexpressed[results1$log2FoldChange > 0.0] = "Upregulated"
count_table1 = table(results1$diffexpressed)
count_table1 = as.data.frame(count_table1)

results2$diffexpressed[results2$log2FoldChange < 0.0] = "Downregulated"
results2$diffexpressed[results2$log2FoldChange > 0.0] = "Upregulated"
count_table2 = table(results2$diffexpressed)
count_table2 = as.data.frame(count_table2)

results3$diffexpressed[results3$log2FoldChange < 0.0] = "Downregulated"
results3$diffexpressed[results3$log2FoldChange >0.0] = "Upregulated"
count_table3 = table(results3$diffexpressed)
count_table3 = as.data.frame(count_table3)

```

```{r updownplot, include=TRUE}

for_barplot = rbind(count_table1, count_table2, count_table3)
for_barplot$DEG = rep(c("Vadadustat Vs. Hypoxia", "Hypoxia Vs. Control", "Vadadustat Vs. Control"), each=2)
colnames(for_barplot) = as.factor(c("Genes", "Counts", "DEG"))

updownplot = ggplot(data=for_barplot, aes(x=DEG, y=Counts, fill=Genes)) +
  geom_bar(stat="identity", position=position_dodge()) +
  geom_text(aes(label=Counts), vjust=0, position=position_dodge(0.9), size=3.5) +
  scale_fill_manual(values=c("red", "gray")) +
  labs(title="Differentially Expressed Genes (DEGs)")
updownplot
ggsave('~/saved_image3.png', plot = last_plot())
```